--- armv7s/include/openssl/opensslconf.h	2017-01-08 02:25:43.000000000 -0800
+++ arm64/include/openssl/opensslconf.h	2017-01-08 03:44:44.000000000 -0800
@@ -158,11 +158,21 @@
  * The following are cipher-specific, but are part of the public API.
  */
 #if !defined(OPENSSL_SYS_UEFI)
-# define BN_LLONG
+# ifdef __LP64__
+#  undef BN_LLONG
+# else
+#  define BN_LLONG
+# endif
 /* Only one for the following should be defined */
-# undef SIXTY_FOUR_BIT_LONG
-# undef SIXTY_FOUR_BIT
-# define THIRTY_TWO_BIT
+# ifdef __LP64__
+#  define SIXTY_FOUR_BIT_LONG
+#  undef SIXTY_FOUR_BIT
+#  undef THIRTY_TWO_BIT
+# else
+#  undef SIXTY_FOUR_BIT_LONG
+#  undef SIXTY_FOUR_BIT
+#  define THIRTY_TWO_BIT
+# endif
 #endif
 
 #define RC4_INT unsigned char
